src/software/ai/hl/stp/tactic/tactic_fsm.h

```diff
@@ -29,42 +31,43 @@ struct TacticUpdate
-        ControlParams control_params;                                                    \
-        TacticUpdate common;                                                             \
+template <class TFsmControlParams>
+class TacticFSM
+{
+   public:
+    using ControlParams = TFsmControlParams;
+    /**
+     * The Update struct is the only event that a tactic FSM should respond to and it is
+     * composed of the following structs:
+     *
+     * TFsmControlParams - uniquely defined parameters for each FSM
+     * TacticUpdate - common struct that contains Robot, World, and SetPrimitiveCallback
+     */
+    struct Update
+    {
+        Update(const TFsmControlParams &control_params, const TacticUpdate &common)
+            : control_params(control_params), common(common)
+        {
+        }
+        TFsmControlParams control_params;
+        TacticUpdate common;
     };
 
-#define DEFINE_TACTIC_DONE_AND_GET_FSM_STATE                                             \
-    bool done() const override                                                           \
-    {                                                                                    \
-        bool is_done = false;                                                            \
-        if (last_execution_robot.has_value())                                            \
-        {                                                                                \
-            is_done = fsm_map.at(last_execution_robot.value())->is(boost::sml::X);       \
-        }                                                                                \
-        return is_done;                                                                  \
-    }                                                                                    \
-                                                                                         \
-    std::string getFSMState() const override                                             \
-    {                                                                                    \
-        std::string state_str = "";                                                      \
-        if (last_execution_robot.has_value())                                            \
-            state_str =                                                                  \
-                getCurrentFullStateName(*fsm_map.at(last_execution_robot.value()));      \
-        return state_str;                                                                \
+    explicit TacticFSM(std::shared_ptr<TbotsProto::AiConfig> ai_config_ptr)
+        : ai_config_ptr(ai_config_ptr)
+    {
     }
 
+   protected:
+    // Former constructors took what they needed from ai_config and stored it locally.
+    // Now, we store ai_config as a pointer and use it to update as needed.
+    // This ensures that everything gets the update as needed.
+    std::shared_ptr<TbotsProto::AiConfig> ai_config_ptr;
```

Comment:
This comment isn't ideal because we need to provide context about what the codebase looked like previously (not necessary, that's what version control is for!) I would suggest changing it to just describe what AiConfig is, e.g.
```suggestion
    // Pointer to shared AiConfig storing dynamic parameters for configuring Tactic behaviour
    std::shared_ptr<TbotsProto::AiConfig> ai_config_ptr;
```